{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1e5900c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nQ1: Explain why we have to use the Exception class while creating a Custom Exception.\\nAnswer:\\nIn Python, the Exception class serves as the base class for all exceptions. When creating a custom exception, \\nit is recommended to inherit from the Exception class to maintain compatibility with the built-in exception hierarchy. \\nThis ensures that your custom exception behaves like a standard exception and allows for consistent handling in try-except\\nblocks. By inheriting from Exception, your custom exception can benefit from the general exception handling mechanisms \\nprovided by the language.\\n'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Q1: Explain why we have to use the Exception class while creating a Custom Exception.\n",
    "Answer:\n",
    "In Python, the Exception class serves as the base class for all exceptions. When creating a custom exception, \n",
    "it is recommended to inherit from the Exception class to maintain compatibility with the built-in exception hierarchy. \n",
    "This ensures that your custom exception behaves like a standard exception and allows for consistent handling in try-except\n",
    "blocks. By inheriting from Exception, your custom exception can benefit from the general exception handling mechanisms \n",
    "provided by the language.\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f94fba31",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'BaseException'>\n",
      "  <class 'Exception'>\n",
      "    <class 'TypeError'>\n",
      "      <class 'decimal.FloatOperation'>\n",
      "      <class 'email.errors.MultipartConversionError'>\n",
      "    <class 'StopAsyncIteration'>\n",
      "    <class 'StopIteration'>\n",
      "    <class 'ImportError'>\n",
      "      <class 'ModuleNotFoundError'>\n",
      "      <class 'zipimport.ZipImportError'>\n",
      "    <class 'OSError'>\n",
      "      <class 'ConnectionError'>\n",
      "        <class 'BrokenPipeError'>\n",
      "        <class 'ConnectionAbortedError'>\n",
      "        <class 'ConnectionRefusedError'>\n",
      "        <class 'ConnectionResetError'>\n",
      "          <class 'http.client.RemoteDisconnected'>\n",
      "      <class 'BlockingIOError'>\n",
      "      <class 'ChildProcessError'>\n",
      "      <class 'FileExistsError'>\n",
      "      <class 'FileNotFoundError'>\n",
      "      <class 'IsADirectoryError'>\n",
      "      <class 'NotADirectoryError'>\n",
      "      <class 'InterruptedError'>\n",
      "        <class 'zmq.error.InterruptedSystemCall'>\n",
      "      <class 'PermissionError'>\n",
      "      <class 'ProcessLookupError'>\n",
      "      <class 'TimeoutError'>\n",
      "      <class 'io.UnsupportedOperation'>\n",
      "      <class 'socket.herror'>\n",
      "      <class 'socket.gaierror'>\n",
      "      <class 'ssl.SSLError'>\n",
      "        <class 'ssl.SSLCertVerificationError'>\n",
      "        <class 'ssl.SSLZeroReturnError'>\n",
      "        <class 'ssl.SSLWantWriteError'>\n",
      "        <class 'ssl.SSLWantReadError'>\n",
      "        <class 'ssl.SSLSyscallError'>\n",
      "        <class 'ssl.SSLEOFError'>\n",
      "      <class 'shutil.Error'>\n",
      "        <class 'shutil.SameFileError'>\n",
      "      <class 'shutil.SpecialFileError'>\n",
      "      <class 'shutil.ExecError'>\n",
      "      <class 'shutil.ReadError'>\n",
      "      <class 'urllib.error.URLError'>\n",
      "        <class 'urllib.error.HTTPError'>\n",
      "        <class 'urllib.error.ContentTooShortError'>\n",
      "      <class 'gzip.BadGzipFile'>\n",
      "    <class 'EOFError'>\n",
      "      <class 'asyncio.exceptions.IncompleteReadError'>\n",
      "    <class 'RuntimeError'>\n",
      "      <class 'RecursionError'>\n",
      "      <class 'NotImplementedError'>\n",
      "        <class 'zmq.error.ZMQVersionError'>\n",
      "        <class 'IPython.core.error.StdinNotImplementedError'>\n",
      "      <class '_frozen_importlib._DeadlockError'>\n",
      "      <class 'threading.BrokenBarrierError'>\n",
      "      <class 'concurrent.futures._base.BrokenExecutor'>\n",
      "        <class 'concurrent.futures.thread.BrokenThreadPool'>\n",
      "      <class 'asyncio.exceptions.SendfileNotAvailableError'>\n",
      "      <class 'pkg_resources.ExtractionError'>\n",
      "      <class '_pydevd_bundle.pydevd_vars.VariableError'>\n",
      "    <class 'NameError'>\n",
      "      <class 'UnboundLocalError'>\n",
      "    <class 'AttributeError'>\n",
      "      <class 'dataclasses.FrozenInstanceError'>\n",
      "    <class 'SyntaxError'>\n",
      "      <class 'IndentationError'>\n",
      "        <class 'TabError'>\n",
      "    <class 'LookupError'>\n",
      "      <class 'IndexError'>\n",
      "      <class 'KeyError'>\n",
      "        <class 'jupyter_client.kernelspec.NoSuchKernel'>\n",
      "        <class 'IPython.terminal.pt_inputhooks.UnknownBackend'>\n",
      "      <class 'encodings.CodecRegistryError'>\n",
      "    <class 'ValueError'>\n",
      "      <class 'UnicodeError'>\n",
      "        <class 'UnicodeEncodeError'>\n",
      "        <class 'UnicodeDecodeError'>\n",
      "        <class 'UnicodeTranslateError'>\n",
      "      <class 'io.UnsupportedOperation'>\n",
      "      <class 'json.decoder.JSONDecodeError'>\n",
      "      <class 'ssl.SSLCertVerificationError'>\n",
      "      <class 'binascii.Error'>\n",
      "      <class '_hashlib.UnsupportedDigestmodError'>\n",
      "      <class 'calendar.IllegalMonthError'>\n",
      "      <class 'calendar.IllegalWeekdayError'>\n",
      "      <class 'dateutil.parser._parser.ParserError'>\n",
      "      <class 'pygments.util.ClassNotFound'>\n",
      "      <class 'IPython.lib.clipboard.ClipboardEmpty'>\n",
      "      <class 'email.errors.MessageDefect'>\n",
      "        <class 'email.errors.NoBoundaryInMultipartDefect'>\n",
      "        <class 'email.errors.StartBoundaryNotFoundDefect'>\n",
      "        <class 'email.errors.CloseBoundaryNotFoundDefect'>\n",
      "        <class 'email.errors.FirstHeaderLineIsContinuationDefect'>\n",
      "        <class 'email.errors.MisplacedEnvelopeHeaderDefect'>\n",
      "        <class 'email.errors.MissingHeaderBodySeparatorDefect'>\n",
      "        <class 'email.errors.MultipartInvariantViolationDefect'>\n",
      "        <class 'email.errors.InvalidMultipartContentTransferEncodingDefect'>\n",
      "        <class 'email.errors.UndecodableBytesDefect'>\n",
      "        <class 'email.errors.InvalidBase64PaddingDefect'>\n",
      "        <class 'email.errors.InvalidBase64CharactersDefect'>\n",
      "        <class 'email.errors.InvalidBase64LengthDefect'>\n",
      "        <class 'email.errors.HeaderDefect'>\n",
      "          <class 'email.errors.InvalidHeaderDefect'>\n",
      "          <class 'email.errors.HeaderMissingRequiredValue'>\n",
      "          <class 'email.errors.NonPrintableDefect'>\n",
      "          <class 'email.errors.ObsoleteHeaderDefect'>\n",
      "          <class 'email.errors.NonASCIILocalPartDefect'>\n",
      "          <class 'email.errors.InvalidDateDefect'>\n",
      "      <class 'IPython.core.magics.code.MacroToEdit'>\n",
      "      <class 'plistlib.InvalidFileException'>\n",
      "      <class 'pkg_resources._vendor.more_itertools.more.UnequalIterablesError'>\n",
      "      <class 'pkg_resources.extern.packaging.version.InvalidVersion'>\n",
      "      <class 'pkg_resources._vendor.packaging._manylinux._ELFFileHeader._InvalidELFFileHeader'>\n",
      "      <class 'pkg_resources.extern.packaging.utils.InvalidWheelFilename'>\n",
      "      <class 'pkg_resources.extern.packaging.utils.InvalidSdistFilename'>\n",
      "      <class 'pkg_resources.extern.packaging.specifiers.InvalidSpecifier'>\n",
      "      <class 'pkg_resources.extern.packaging.markers.InvalidMarker'>\n",
      "      <class 'pkg_resources.extern.packaging.markers.UndefinedComparison'>\n",
      "      <class 'pkg_resources.extern.packaging.markers.UndefinedEnvironmentName'>\n",
      "      <class 'pkg_resources.extern.packaging.requirements.InvalidRequirement'>\n",
      "        <class 'pkg_resources.RequirementParseError'>\n",
      "      <class 'packaging.version.InvalidVersion'>\n",
      "    <class 'AssertionError'>\n",
      "    <class 'ArithmeticError'>\n",
      "      <class 'FloatingPointError'>\n",
      "      <class 'OverflowError'>\n",
      "      <class 'ZeroDivisionError'>\n",
      "        <class 'decimal.DivisionByZero'>\n",
      "        <class 'decimal.DivisionUndefined'>\n",
      "      <class 'decimal.DecimalException'>\n",
      "        <class 'decimal.Clamped'>\n",
      "        <class 'decimal.Rounded'>\n",
      "          <class 'decimal.Underflow'>\n",
      "          <class 'decimal.Overflow'>\n",
      "        <class 'decimal.Inexact'>\n",
      "          <class 'decimal.Underflow'>\n",
      "          <class 'decimal.Overflow'>\n",
      "        <class 'decimal.Subnormal'>\n",
      "          <class 'decimal.Underflow'>\n",
      "        <class 'decimal.DivisionByZero'>\n",
      "        <class 'decimal.FloatOperation'>\n",
      "        <class 'decimal.InvalidOperation'>\n",
      "          <class 'decimal.ConversionSyntax'>\n",
      "          <class 'decimal.DivisionImpossible'>\n",
      "          <class 'decimal.DivisionUndefined'>\n",
      "          <class 'decimal.InvalidContext'>\n",
      "    <class 'SystemError'>\n",
      "      <class 'encodings.CodecRegistryError'>\n",
      "    <class 'ReferenceError'>\n",
      "    <class 'MemoryError'>\n",
      "    <class 'BufferError'>\n",
      "    <class 'Warning'>\n",
      "      <class 'UserWarning'>\n",
      "        <class 'getpass.GetPassWarning'>\n",
      "        <class 'IPython.core.formatters.FormatterWarning'>\n",
      "      <class 'EncodingWarning'>\n",
      "      <class 'DeprecationWarning'>\n",
      "        <class 'IPython.core.interactiveshell.ProvisionalWarning'>\n",
      "      <class 'PendingDeprecationWarning'>\n",
      "      <class 'SyntaxWarning'>\n",
      "      <class 'RuntimeWarning'>\n",
      "        <class 'zmq.asyncio.ProactorSelectorThreadWarning'>\n",
      "        <class 'dateutil.parser._parser.UnknownTimezoneWarning'>\n",
      "        <class 'pkg_resources.PEP440Warning'>\n",
      "      <class 'FutureWarning'>\n",
      "        <class 'IPython.core.completer.ProvisionalCompleterWarning'>\n",
      "      <class 'ImportWarning'>\n",
      "      <class 'UnicodeWarning'>\n",
      "      <class 'BytesWarning'>\n",
      "      <class 'ResourceWarning'>\n",
      "      <class 'dateutil.tz.DeprecatedTzFormatWarning'>\n",
      "      <class 'pkg_resources.PkgResourcesDeprecationWarning'>\n",
      "    <class 'warnings._OptionError'>\n",
      "    <class 'runpy._Error'>\n",
      "    <class 're.error'>\n",
      "    <class 'sre_parse.Verbose'>\n",
      "    <class 'locale.Error'>\n",
      "    <class 'subprocess.SubprocessError'>\n",
      "      <class 'subprocess.CalledProcessError'>\n",
      "      <class 'subprocess.TimeoutExpired'>\n",
      "    <class 'tokenize.TokenError'>\n",
      "    <class 'tokenize.StopTokenizing'>\n",
      "    <class 'inspect.ClassFoundException'>\n",
      "    <class 'inspect.EndOfBlock'>\n",
      "    <class 'traitlets.traitlets.TraitError'>\n",
      "    <class 'copy.Error'>\n",
      "    <class 'concurrent.futures._base.Error'>\n",
      "      <class 'concurrent.futures._base.CancelledError'>\n",
      "      <class 'concurrent.futures._base.TimeoutError'>\n",
      "      <class 'concurrent.futures._base.InvalidStateError'>\n",
      "    <class 'socket._GiveupOnSendfile'>\n",
      "    <class 'struct.error'>\n",
      "    <class 'binascii.Incomplete'>\n",
      "    <class 'asyncio.exceptions.TimeoutError'>\n",
      "    <class 'asyncio.exceptions.InvalidStateError'>\n",
      "    <class 'asyncio.exceptions.LimitOverrunError'>\n",
      "    <class 'asyncio.queues.QueueEmpty'>\n",
      "    <class 'asyncio.queues.QueueFull'>\n",
      "    <class 'zlib.error'>\n",
      "    <class '_lzma.LZMAError'>\n",
      "    <class 'shutil.RegistryError'>\n",
      "    <class 'shutil._GiveupOnFastCopy'>\n",
      "    <class '_queue.Empty'>\n",
      "    <class 'queue.Full'>\n",
      "    <class 'zmq.error.ZMQBaseError'>\n",
      "      <class 'zmq.error.ZMQError'>\n",
      "        <class 'zmq.error.ContextTerminated'>\n",
      "        <class 'zmq.error.Again'>\n",
      "        <class 'zmq.error.InterruptedSystemCall'>\n",
      "      <class 'zmq.error.ZMQBindError'>\n",
      "      <class 'zmq.error.NotDone'>\n",
      "    <class '_pickle.PickleError'>\n",
      "      <class '_pickle.PicklingError'>\n",
      "      <class '_pickle.UnpicklingError'>\n",
      "    <class 'pickle._Stop'>\n",
      "    <class 'argparse.ArgumentError'>\n",
      "    <class 'argparse.ArgumentTypeError'>\n",
      "    <class 'traitlets.config.loader.ConfigError'>\n",
      "      <class 'traitlets.config.loader.ConfigLoaderError'>\n",
      "        <class 'traitlets.config.loader.ArgumentError'>\n",
      "      <class 'traitlets.config.loader.ConfigFileNotFound'>\n",
      "    <class 'traitlets.config.configurable.ConfigurableError'>\n",
      "      <class 'traitlets.config.configurable.MultipleInstanceError'>\n",
      "    <class 'traitlets.config.application.ApplicationError'>\n",
      "    <class 'tornado.util.TimeoutError'>\n",
      "    <class 'ctypes.ArgumentError'>\n",
      "    <class '_ctypes.COMError'>\n",
      "    <class 'tornado.concurrent.ReturnValueIgnoredError'>\n",
      "    <class 'tornado.gen.KeyReuseError'>\n",
      "    <class 'tornado.gen.UnknownKeyError'>\n",
      "    <class 'tornado.gen.LeakedCallbackError'>\n",
      "    <class 'tornado.gen.BadYieldError'>\n",
      "    <class 'tornado.gen.ReturnValueIgnoredError'>\n",
      "    <class 'tornado.gen.Return'>\n",
      "    <class 'jupyter_client.channels.InvalidPortNumber'>\n",
      "    <class 'jupyter_client.localinterfaces.NoIPAddresses'>\n",
      "    <class 'zipfile.BadZipFile'>\n",
      "    <class 'zipfile.LargeZipFile'>\n",
      "    <class 'configparser.Error'>\n",
      "      <class 'configparser.NoSectionError'>\n",
      "      <class 'configparser.DuplicateSectionError'>\n",
      "      <class 'configparser.DuplicateOptionError'>\n",
      "      <class 'configparser.NoOptionError'>\n",
      "      <class 'configparser.InterpolationError'>\n",
      "        <class 'configparser.InterpolationMissingOptionError'>\n",
      "        <class 'configparser.InterpolationSyntaxError'>\n",
      "        <class 'configparser.InterpolationDepthError'>\n",
      "      <class 'configparser.ParsingError'>\n",
      "        <class 'configparser.MissingSectionHeaderError'>\n",
      "    <class 'entrypoints.BadEntryPoint'>\n",
      "    <class 'entrypoints.NoSuchEntryPoint'>\n",
      "    <class 'jupyter_client.multikernelmanager.DuplicateKernelError'>\n",
      "    <class 'pydoc.ErrorDuringImport'>\n",
      "    <class 'executing.executing.NotOneValueFound'>\n",
      "    <class 'pure_eval.utils.CannotEval'>\n",
      "    <class 'pygments.util.OptionError'>\n",
      "    <class 'bdb.BdbQuit'>\n",
      "    <class 'pdb.Restart'>\n",
      "    <class 'IPython.utils.process.FindCmdError'>\n",
      "    <class 'IPython.utils.path.HomeDirError'>\n",
      "    <class 'IPython.core.profiledir.ProfileDirError'>\n",
      "    <class 'IPython.core.error.IPythonCoreError'>\n",
      "      <class 'IPython.core.error.TryNext'>\n",
      "      <class 'IPython.core.error.UsageError'>\n",
      "      <class 'IPython.core.error.StdinNotImplementedError'>\n",
      "    <class 'IPython.core.error.InputRejected'>\n",
      "    <class 'getopt.GetoptError'>\n",
      "    <class 'pygments.filters.ErrorToken'>\n",
      "    <class 'IPython.core.prefilter.PrefilterError'>\n",
      "    <class 'IPython.core.alias.AliasError'>\n",
      "      <class 'IPython.core.alias.InvalidAliasError'>\n",
      "    <class 'sqlite3.Error'>\n",
      "      <class 'sqlite3.InterfaceError'>\n",
      "      <class 'sqlite3.DatabaseError'>\n",
      "        <class 'sqlite3.InternalError'>\n",
      "        <class 'sqlite3.OperationalError'>\n",
      "        <class 'sqlite3.ProgrammingError'>\n",
      "        <class 'sqlite3.IntegrityError'>\n",
      "        <class 'sqlite3.DataError'>\n",
      "        <class 'sqlite3.NotSupportedError'>\n",
      "    <class 'sqlite3.Warning'>\n",
      "    <class 'IPython.core.interactiveshell.SpaceInInput'>\n",
      "    <class 'xml.dom.DOMException'>\n",
      "      <class 'xml.dom.IndexSizeErr'>\n",
      "      <class 'xml.dom.DomstringSizeErr'>\n",
      "      <class 'xml.dom.HierarchyRequestErr'>\n",
      "      <class 'xml.dom.WrongDocumentErr'>\n",
      "      <class 'xml.dom.InvalidCharacterErr'>\n",
      "      <class 'xml.dom.NoDataAllowedErr'>\n",
      "      <class 'xml.dom.NoModificationAllowedErr'>\n",
      "      <class 'xml.dom.NotFoundErr'>\n",
      "      <class 'xml.dom.NotSupportedErr'>\n",
      "      <class 'xml.dom.InuseAttributeErr'>\n",
      "      <class 'xml.dom.InvalidStateErr'>\n",
      "      <class 'xml.dom.SyntaxErr'>\n",
      "      <class 'xml.dom.InvalidModificationErr'>\n",
      "      <class 'xml.dom.NamespaceErr'>\n",
      "      <class 'xml.dom.InvalidAccessErr'>\n",
      "      <class 'xml.dom.ValidationErr'>\n",
      "    <class 'prompt_toolkit.validation.ValidationError'>\n",
      "    <class 'prompt_toolkit.buffer.EditReadOnlyBuffer'>\n",
      "    <class 'prompt_toolkit.buffer._Retry'>\n",
      "    <class 'prompt_toolkit.layout.layout.InvalidLayoutError'>\n",
      "    <class 'prompt_toolkit.renderer.HeightIsUnknownError'>\n",
      "    <class 'IPython.core.guarded_eval.GuardRejection'>\n",
      "    <class 'parso.parser.ParserSyntaxError'>\n",
      "    <class 'parso.parser.InternalParseError'>\n",
      "    <class 'parso.python.diff._PositionUpdatingFinished'>\n",
      "    <class 'jedi.inference.helpers.SimpleGetItemNotFound'>\n",
      "    <class 'jedi.inference.utils.UncaughtAttributeError'>\n",
      "    <class 'jedi.inference.base_value.HasNoContext'>\n",
      "    <class 'jedi.inference.arguments.ParamIssue'>\n",
      "    <class 'jedi.api.exceptions._JediError'>\n",
      "      <class 'jedi.api.exceptions.InternalError'>\n",
      "      <class 'jedi.api.exceptions.WrongVersion'>\n",
      "      <class 'jedi.api.exceptions.RefactoringError'>\n",
      "    <class 'jedi.api.helpers.OnErrorLeaf'>\n",
      "    <class 'jedi.api.environment.InvalidPythonEnvironment'>\n",
      "    <class 'email.errors.MessageError'>\n",
      "      <class 'email.errors.MessageParseError'>\n",
      "        <class 'email.errors.HeaderParseError'>\n",
      "        <class 'email.errors.BoundaryError'>\n",
      "      <class 'email.errors.MultipartConversionError'>\n",
      "      <class 'email.errors.CharsetError'>\n",
      "    <class 'uu.Error'>\n",
      "    <class 'http.client.HTTPException'>\n",
      "      <class 'http.client.NotConnected'>\n",
      "      <class 'http.client.InvalidURL'>\n",
      "      <class 'http.client.UnknownProtocol'>\n",
      "      <class 'http.client.UnknownTransferEncoding'>\n",
      "      <class 'http.client.UnimplementedFileMode'>\n",
      "      <class 'http.client.IncompleteRead'>\n",
      "      <class 'http.client.ImproperConnectionState'>\n",
      "        <class 'http.client.CannotSendRequest'>\n",
      "        <class 'http.client.CannotSendHeader'>\n",
      "        <class 'http.client.ResponseNotReady'>\n",
      "      <class 'http.client.BadStatusLine'>\n",
      "        <class 'http.client.RemoteDisconnected'>\n",
      "      <class 'http.client.LineTooLong'>\n",
      "    <class 'IPython.core.magics.code.InteractivelyDefined'>\n",
      "    <class 'IPython.terminal.embed.KillEmbedded'>\n",
      "    <class 'psutil.Error'>\n",
      "      <class 'psutil.NoSuchProcess'>\n",
      "        <class 'psutil.ZombieProcess'>\n",
      "      <class 'psutil.AccessDenied'>\n",
      "      <class 'psutil.TimeoutExpired'>\n",
      "    <class '_psutil_windows.Error'>\n",
      "    <class '_psutil_windows.TimeoutExpired'>\n",
      "    <class '_psutil_windows.TimeoutAbandoned'>\n",
      "    <class 'tornado.queues.QueueEmpty'>\n",
      "    <class 'tornado.queues.QueueFull'>\n",
      "    <class 'xml.parsers.expat.ExpatError'>\n",
      "    <class 'xmlrpc.client.Error'>\n",
      "      <class 'xmlrpc.client.ProtocolError'>\n",
      "      <class 'xmlrpc.client.ResponseError'>\n",
      "      <class 'xmlrpc.client.Fault'>\n",
      "    <class 'pywintypes.error'>\n",
      "    <class 'pywintypes.com_error'>\n",
      "    <class 'pythoncom.internal_error'>\n",
      "    <class 'pkg_resources._vendor.pyparsing.exceptions.ParseBaseException'>\n",
      "      <class 'pkg_resources._vendor.pyparsing.exceptions.ParseException'>\n",
      "      <class 'pkg_resources._vendor.pyparsing.exceptions.ParseFatalException'>\n",
      "        <class 'pkg_resources._vendor.pyparsing.exceptions.ParseSyntaxException'>\n",
      "    <class 'pkg_resources._vendor.pyparsing.exceptions.RecursiveGrammarException'>\n",
      "    <class 'pkg_resources.ResolutionError'>\n",
      "      <class 'pkg_resources.VersionConflict'>\n",
      "        <class 'pkg_resources.ContextualVersionConflict'>\n",
      "      <class 'pkg_resources.DistributionNotFound'>\n",
      "      <class 'pkg_resources.UnknownExtra'>\n",
      "    <class '_pydevd_bundle.pydevd_resolver.UnableToResolveVariableException'>\n",
      "    <class '_pydev_bundle.pydev_monkey.InvalidTypeInArgsException'>\n",
      "    <class '__main__.CustomError'>\n",
      "    <class '__main__.MyCustomException'>\n",
      "  <class 'GeneratorExit'>\n",
      "  <class 'SystemExit'>\n",
      "  <class 'KeyboardInterrupt'>\n",
      "  <class 'asyncio.exceptions.CancelledError'>\n",
      "  <class '_pydev_imps._pydev_saved_modules.DebuggerInitializationError'>\n",
      "  <class 'pkg_resources._vendor.more_itertools.more.AbortThread'>\n"
     ]
    }
   ],
   "source": [
    "'''Q2: Write a Python program to print Python Exception Hierarchy.'''\n",
    "\n",
    "# Program :\n",
    "\n",
    "def print_exception_hierarchy(exception_class, indent=0):\n",
    "    print(\"  \" * indent + str(exception_class))\n",
    "    for subclass in exception_class.__subclasses__():\n",
    "        print_exception_hierarchy(subclass, indent + 1)\n",
    "\n",
    "print_exception_hierarchy(BaseException)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d56b2c0b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Caught an exception: division by zero\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "Q3: What errors are defined in the ArithmeticError class? Explain any two with an example.\n",
    "Ans:\n",
    "The ArithmeticError class is a base class for errors that occur during arithmetic operations. Two common errors under \n",
    "ArithmeticError are\n",
    "'''\n",
    "# ZeroDivisionError - Raised when division or modulo by zero is encountered.\n",
    "try:\n",
    "    result = 10 / 0  # This will raise a ZeroDivisionError\n",
    "except ZeroDivisionError as e:\n",
    "    print(f\"Caught an exception: {e}\")\n",
    "    \n",
    "#OverflowError - Raised when the result of an arithmetic operation is too large to be represented\n",
    "import sys\n",
    "\n",
    "try:\n",
    "    result = sys.maxsize * 2  # This will raise an OverflowError on most systems\n",
    "except OverflowError as e:\n",
    "    print(f\"Caught an exception: {e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "798b790b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Caught an exception: 'c'\n",
      "Caught an exception: list index out of range\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "Q4: Why LookupError class is used? Explain with an example KeyError and IndexError.\n",
    "Answer:\n",
    "The LookupError class is a base class for errors that occur when a key or index used to access a mapping or sequence \n",
    "is invalid. Two common errors under LookupError are:\n",
    "'''\n",
    "\n",
    "#KeyError - Raised when a dictionary key is not found.\n",
    "my_dict = {'a': 1, 'b': 2}\n",
    "try:\n",
    "    value = my_dict['c']  # This will raise a KeyError\n",
    "except KeyError as e:\n",
    "    print(f\"Caught an exception: {e}\")\n",
    "\n",
    "#IndexError - Raised when trying to access an index that is out of range.\n",
    "my_list = [1, 2, 3]\n",
    "try:\n",
    "    value = my_list[5]  # This will raise an IndexError\n",
    "except IndexError as e:\n",
    "    print(f\"Caught an exception: {e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9eaa354d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Caught an exception: No module named 'non_existing_module'\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "Q5: Explain ImportError. What is ModuleNotFoundError?\n",
    "Answer:\n",
    "ImportError is a base class for exceptions related to importing modules. ModuleNotFoundError is a subclass of ImportError \n",
    "that specifically indicates that a module cannot be found.\n",
    "'''\n",
    "\n",
    "# Program :\n",
    "\n",
    "try:\n",
    "    import non_existing_module  # This will raise a ModuleNotFoundError\n",
    "except ModuleNotFoundError as e:\n",
    "    print(f\"Caught an exception: {e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b42fe3e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nQ6: List down some best practices for exception handling in Python.\\nAnswer:\\n\\n1. Use Specific Exceptions: Catch specific exceptions rather than using a broad except clause. It helps in understanding and \\nhandling specific issues.\\n\\n2. Avoid Bare except: Avoid using a bare except clause as it catches all exceptions and may hide bugs. Be specific about \\nthe exceptions you want to catch.\\n\\n3. Use finally for Cleanup: Use finally for cleanup code that should be executed regardless of whether an exception \\noccurred or not.\\n\\n4. Keep Try Blocks Small: Keep the code inside try blocks minimal to focus on the specific code that might raise an \\nexception.\\n\\n5. Handle Exceptions Locally: Handle exceptions as close to the source as possible. Avoid catching exceptions too broadly.\\n\\n6. Log Exceptions: Use logging to record exceptions, including the traceback. It helps in debugging and analyzing issues.\\n\\n7. Avoid Silent Failures: Avoid catching exceptions without taking any action. Silent failures make it challenging to \\nidentify and fix problems.\\n\\n8. Custom Exception Messages: Provide informative and meaningful error messages when raising or catching exceptions. \\nIt aids in understanding the cause of the issue.\\n\\n9. Use with for Resource Management: When dealing with resources like files or network connections, use with statements \\nto ensure proper cleanup even in the presence of exceptions.\\n\\n10. Test Exception Handling: Include testing scenarios that involve exceptions to ensure your exception handling mechanisms\\nwork as expected.\\n\\n'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Q6: List down some best practices for exception handling in Python.\n",
    "Answer:\n",
    "\n",
    "1. Use Specific Exceptions: Catch specific exceptions rather than using a broad except clause. It helps in understanding and \n",
    "handling specific issues.\n",
    "\n",
    "2. Avoid Bare except: Avoid using a bare except clause as it catches all exceptions and may hide bugs. Be specific about \n",
    "the exceptions you want to catch.\n",
    "\n",
    "3. Use finally for Cleanup: Use finally for cleanup code that should be executed regardless of whether an exception \n",
    "occurred or not.\n",
    "\n",
    "4. Keep Try Blocks Small: Keep the code inside try blocks minimal to focus on the specific code that might raise an \n",
    "exception.\n",
    "\n",
    "5. Handle Exceptions Locally: Handle exceptions as close to the source as possible. Avoid catching exceptions too broadly.\n",
    "\n",
    "6. Log Exceptions: Use logging to record exceptions, including the traceback. It helps in debugging and analyzing issues.\n",
    "\n",
    "7. Avoid Silent Failures: Avoid catching exceptions without taking any action. Silent failures make it challenging to \n",
    "identify and fix problems.\n",
    "\n",
    "8. Custom Exception Messages: Provide informative and meaningful error messages when raising or catching exceptions. \n",
    "It aids in understanding the cause of the issue.\n",
    "\n",
    "9. Use with for Resource Management: When dealing with resources like files or network connections, use with statements \n",
    "to ensure proper cleanup even in the presence of exceptions.\n",
    "\n",
    "10. Test Exception Handling: Include testing scenarios that involve exceptions to ensure your exception handling mechanisms\n",
    "work as expected.\n",
    "\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "520d73bd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
